Unhandled web process message 'WebPage_RegisterURLSchemeHandler' on every MiniBrowser startup

basuke opened the pull request:
#### fe92a47ca059dd3225381831387208483c2efeba
<pre>
Unhandled web process message &apos;WebPage_RegisterURLSchemeHandler&apos; on every MiniBrowser startup
<a href="https://bugs.webkit.org/show_bug.cgi?id=290246">https://bugs.webkit.org/show_bug.cgi?id=290246</a>
<a href="https://rdar.apple.com/147630372">rdar://147630372</a>

Reviewed by Chris Dumez.

When registering URLSchemeHandler, it sends the message to the webprocess to report the
registration of the handler. But when it happens in the very beginning of the WebPage
creation, the timing is very unstable.

To solve this, fix the registration at the creation of WebPageProxy instance and make
sending message as the conditional. If the process is not attached to WebProcess, it
won&apos;t send the registration information to the process. Even if it won&apos;t be sent, they
will be passed to WebProcess via WebPageCreationParameters.

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::m_pageForTesting):
(WebKit::WebPageProxy::didAttachToRunningProcess):
(WebKit::WebPageProxy::setURLSchemeHandlerForScheme):

Canonical link: <a href="https://commits.webkit.org/292615@main">https://commits.webkit.org/292615@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b65fa4ece24ffa16771827f01fa0a00d0c1a4dbc

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/96342 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/15956 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/5934 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/101409 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/46861 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/16252 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/24389 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/73431 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/30660 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/99345 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/12198 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/87032 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/53768 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/11954 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/46188 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/82063 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/4897 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/103437 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/23409 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/17041 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/82473 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/23660 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/83057 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/81850 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20597 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/26489 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/3927 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/16789 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/23372 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/28527 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/23031 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/26511 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/24772 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -14183,7 +14182,8 @@ void WebPageProxy::setURLSchemeHandlerForScheme(Ref<WebURLSchemeHandler>&& handl
 
     WebCore::LegacySchemeRegistry::registerURLSchemeAsHandledBySchemeHandler(scheme);
-    send(Messages::WebPage::RegisterURLSchemeHandler(handlerIdentifier, canonicalizedScheme.value()));
+    if (hasRunningProcess())
```
If we stop sending this message, won't it break URLSchemeHandlers in some cases?  Couldn't we instead move the call to WKWebView._initializeWithConfiguration with the other calls to setURLSchemeHandlerForScheme?

cdumez replied:
The commit log mentions that they will get sent via `WebPageCreationParameters` later on.

basuke replied:
I read the code flow again and here is my understanding.
- on constructor of WebPageProxy, either WebProcess is connected or not.
  - if connected, registered scheme handler is sent via `WebPage::RegisterURLSchemeHandler` immediately.
  - if not, it just stored in `m_urlSchemeHandlersByScheme`.
    - These values will be passed to WebProcess via `WebProcess::CreateWebPage`.

In `launchProcess()`,
- `m_hasRunningProcess` set to true
- `finishAttachingToWebProcess()` is called and `WebProcess::CreateWebPage` is invoked in it.

So I believe there's no chance to miss delivering scheme handler information even when they've registered in between WebPageProxy creation and WebProcess launch.

achristensen07 replied:
ok

basuke committed:
Unhandled web process message 'WebPage_RegisterURLSchemeHandler' on every MiniBrowser startup
https://bugs.webkit.org/show_bug.cgi?id=290246
rdar://147630372

Reviewed by Chris Dumez.

When registering URLSchemeHandler, it sends the message to the webprocess to report the
registration of the handler. But when it happens in the very beginning of the WebPage
creation, the timing is very unstable.

To solve this, fix the registration at the creation of WebPageProxy instance and make
sending message as the conditional. If the process is not attached to WebProcess, it
won't send the registration information to the process. Even if it won't be sent, they
will be passed to WebProcess via WebPageCreationParameters.

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::m_pageForTesting):
(WebKit::WebPageProxy::didAttachToRunningProcess):
(WebKit::WebPageProxy::setURLSchemeHandlerForScheme):

Canonical link: https://commits.webkit.org/292615@main
